### PR TITLE
`crystal init`: Improve transformation of project name with hyphens

### PR DESCRIPTION
**Pull request by Kanezoh**
issue #11107 

when executing `crystal init app crystal-proj-1`

before

~~~
module Crystal::Proj::1
~~~

after

~~~
module Crystal::Proj1
~~~

### CODE REVIEW

File: src/compiler/crystal/tools/init.cr [hunk 1]
```diff
@@ -207,7 +207,7 @@ module Crystal
       end
 
       def module_name
-        config.name.split('-').map(&.camelcase).join("::")
+        config.name.gsub(/-([0-9]+)/, "\\1").split('-').map(&.camelcase).join("::")
```
Comment (Sija) [inline src/compiler/crystal/tools/init.cr:210]:
```suggestion
        config.name.split('-').compact_map { |name| name.camelcase if name[0]?.try(&.ascii_letter?) }.join("::")
```

Comment (Kanezoh) [reply src/compiler/crystal/tools/init.cr:210]: Thank you for suggestion!
It returns `Crystal::Proj`, so `1` is ignored. Is it OK?

Comment (Sija) [reply src/compiler/crystal/tools/init.cr:210]: You're right, I've updated the suggestion.

**Comment by Kanezoh** [comment]
Thanks. It seems more preciese, and certainly returns `Crystal::Proj1`.
I've accepted.

**Review by straight-shoota — COMMENTED** [review]
With this method growing more complex, I'd like to add some specs for it. Can you take care of that?

File: src/compiler/crystal/tools/init.cr [hunk 2]
```diff
@@ -207,7 +207,7 @@ module Crystal
       end
 
       def module_name
-        config.name.split('-').map(&.camelcase).join("::")
+        config.name.gsub(/[-_]([^a-z])/i, "\\1").split('-').compact_map { |name| name.camelcase if name[0]?.try(&.ascii_letter?) }.join("::")
```
[the remark this answers is not in the document]
Comment (Sija) [reply src/compiler/crystal/tools/init.cr:210]: The valid path part must start with a letter, so if that's not the case we skip it altogether.

**Comment by Kanezoh** [comment]
I'm not sure it's enough, but added a test.

**Review by straight-shoota — APPROVED** [review]
I was however thinking about unit tests for this specific method, specifying the expected output for a number of edge cases.
But don't feel pressed to do that. The PR is good enough for me.

**Comment by Kanezoh** [comment]
I also thought testing `module_name` method only is best, but couldn't think of good way to do it.
Especially, I have difficulty in making `View` instance for test.
I only came up with the idea like below.

in `init_spec.cr`
~~~

  class Crystal::Init::TestView < Crystal::Init::View
    def path
      "test_path"
    end
  end

  config = Crystal::Init::Config.new
  config.name = "proj-1"
  test_view = Crystal::Init::TestView.new(config)
  puts test_view.module_name # => proj1
~~~

`View` is abstract class, so I have to define `TestView` class. Is defining new class inside the test OK? I don't think it's good way, so decide to current test. I want idea to test `module_name` method with better way.

I'm new in Crystal, so sorry if I made a mistake.

**Comment by straight-shoota** [comment]
Yeah, it's not trivial to setup tests for this class.

Introducing a `TestView` subclass is probably fine. Alternatively, you could extract the implementation of `#module_name` to a class method which can be called directly with a `name` parameter.

**Comment by Kanezoh** [comment]
Thank you for your idea!
Do you mean like this?

~~~
  class TestView
    def self.module_name(name)
      name.gsub(/[-_]([^a-z])/i, "\\1").split('-').compact_map { |name| name.camelcase if name[0]?.try(&.ascii_letter?) }.join("::")
    end
  end

  puts TestView.module_name("proj-1") # => Proj1
~~~

**Comment by straight-shoota** [comment]
No, I mean changing the implementation of `View#module_name` to delegate to a dedicated class method which can then be tested in isolation:

```cr
abstract class Crystal::Init::View
  def module_name
    View.module_name config.name
  end

  def self.module_name(name)
    name.gsub(/[-_]([^a-z])/i, "\\1").split('-').compact_map { |name| name.camelcase if name[0]?.try(&.ascii_letter?) }.join("::")
  end
end

it do
  Crystal::Init::View.module_name("proj-1").should eq "Proj1"
end
```

**Comment by Kanezoh** [comment]
Oh, now I understand. try to implement as so. Thank you!

**Comment by Kanezoh** [comment]
I've defined module_name as a class method, and add tests for module_name method.

File: spec/compiler/crystal/tools/init_spec.cr [hunk 3]
```diff
@@ -388,4 +393,16 @@ module Crystal
...
   end
+
+  describe "View#module_name" do
+    it "namespace is divided by hyphen" do
+      Crystal::Init::View.module_name("my-proj-name").should eq "My::Proj::Name"
+    end
+    it "hyphen follwed by non-ascii letter is replaced by its character" do
+      Crystal::Init::View.module_name("my-proj-1").should eq "My::Proj1"
+    end
+    it "unserscore is ignored" do
```
Comment (straight-shoota) [inline spec/compiler/crystal/tools/init_spec.cr:404]:
```suggestion
    it "underscore is ignored" do
```

**Review by straight-shoota — APPROVED** [review]

**Comment by Kanezoh** [comment]
sorry, accepted fixing typo.

File: src/compiler/crystal/tools/init.cr [hunk 4]
```diff
@@ -207,7 +207,11 @@ module Crystal
       end
 
       def module_name
-        config.name.gsub(/[-_]([^a-z])/i, "\\1").split('-').compact_map { |name| name.camelcase if name[0]?.try(&.ascii_letter?) }.join("::")
+        View.module_name(config.name)
+      end
+
+      def self.module_name(name)
+        name.gsub(/[-_]([^a-z])/i, "\\1").split('-').compact_map { |name| name.camelcase if name[0]?.try(&.ascii_letter?) }.join("::")
```
Comment (Sija) [inline src/compiler/crystal/tools/init.cr:214]: Might be a bit more readable this way.

```suggestion
        name
          .gsub(/[-_]([^a-z])/i, "\\1")
          .split('-')
          .compact_map do |name| 
            name.camelcase if name[0]?.try(&.ascii_letter?)
          end
          .join("::")
```

Comment (Kanezoh) [reply src/compiler/crystal/tools/init.cr:214]: looks fine!

**Review by sdogruyol — APPROVED** [review]
Thank you @Kanezoh 🙏
